### PR TITLE
[SYCL-MLIR]: Memory access analysis (part6)

### PR DESCRIPTION
**Pull request by etiotto**
Extend the memory access analysis to produce an offset vector in addition to an access matrix.

### CODE REVIEW

File: polygeist/lib/Dialect/Polygeist/Analysis/MemoryAccessAnalysis.cpp [hunk 1]
```diff
@@ -124,77 +124,95 @@ static bool usesValue(Operation *op, Value val) {
...
+            typename = std::enable_if_t<llvm::is_one_of<U, T, Value>::value>>
+  U get() const {
+    assert(is<U>() && "Incorrect type");
+    return std::get<U>(data);
   }
 
-  /// Determine whether this class holds the value one.
-  bool isOne(DataFlowSolver &solver) const {
-    Value val = isMultiplier() ? getMultiplier().getValue() : getValue();
+  /// Determine whether this class holds the value \p k.
```
Comment (whitneywhtsang) [inline polygeist/lib/Dialect/Polygeist/Analysis/MemoryAccessAnalysis.cpp:196]:
```suggestion
  /// Determine whether this class holds the value \p constant.
```

**Review by sommerlukas — APPROVED** [review]

File: polygeist/lib/Dialect/Polygeist/Analysis/MemoryAccessAnalysis.cpp [hunk 2]
```diff
@@ -338,12 +357,183 @@ static ValueOrMultiplier getMultiplier(const Value expr, const Value factor,
       .Default([&](auto) { return expr; });
 }
 
+static Value removeConversions(Value expr) {
```
Comment (whitneywhtsang) [inline polygeist/lib/Dialect/Polygeist/Analysis/MemoryAccessAnalysis.cpp:360]: this static function is not used.

File: polygeist/lib/Dialect/Polygeist/Analysis/MemoryAccessAnalysis.cpp [hunk 3]
```diff
@@ -338,12 +357,183 @@ static ValueOrMultiplier getMultiplier(const Value expr, const Value factor,
...
+          return Value();
+        };
+
+        LLVM_DEBUG(llvm::dbgs() << "expr: " << expr << "\n");
+
+        return visitBinaryOp(mulOp, loopAndThreadVars, solver, getOffset,
+                             computeResult);
+      })
+      .Case<CastOpInterface>([&](auto) {
+        return getOffset(op->getOperand(0), loopAndThreadVars, solver);
```
Comment (whitneywhtsang) [inline polygeist/lib/Dialect/Polygeist/Analysis/MemoryAccessAnalysis.cpp:526]: return `Value()` or `expr` when number of operands != 1, as operation with CastOpInterface can have N inputs and N outputs.

File: polygeist/include/mlir/Dialect/Polygeist/Analysis/MemoryAccessAnalysis.h [hunk 4]
```diff
@@ -389,23 +387,17 @@ inline raw_ostream &operator<<(raw_ostream &os, const OffsetVector &vector) {
...
-      typename T = OpTy,
-      typename = std::enable_if_t<llvm::is_one_of<T, affine::AffineLoadOp,
-                                                  affine::AffineStoreOp>::value,
-                                  bool>>
-  MemoryAccess(T accessOp, MemoryAccessMatrix &&matrix, OffsetVector &&offsets)
-      : accessOp(accessOp), matrix(std::move(matrix)),
-        offsets(std::move(offsets)) {}
-
-  OpTy getAccessOp() const { return accessOp; }
+  MemoryAccess() = default;
```
Comment (whitneywhtsang) [inline polygeist/include/mlir/Dialect/Polygeist/Analysis/MemoryAccessAnalysis.h:395]: When do we want to create MemoryAccess without the access matrix and offset vector?

Comment (etiotto) [reply polygeist/include/mlir/Dialect/Polygeist/Analysis/MemoryAccessAnalysis.h:395]: The default constructor is required in order to create the DenseMap at line 485 (DenseMap<Operation *, MemoryAccess> accessMap).

File: polygeist/lib/Dialect/Polygeist/Analysis/MemoryAccessAnalysis.cpp (hunk 1 above)
Comment (etiotto) [reply polygeist/lib/Dialect/Polygeist/Analysis/MemoryAccessAnalysis.cpp:196]: done

File: polygeist/lib/Dialect/Polygeist/Analysis/MemoryAccessAnalysis.cpp (hunk 2 above)
Comment (etiotto) [reply polygeist/lib/Dialect/Polygeist/Analysis/MemoryAccessAnalysis.cpp:360]: Removed.

File: polygeist/lib/Dialect/Polygeist/Analysis/MemoryAccessAnalysis.cpp (hunk 3 above)
Comment (etiotto) [reply polygeist/lib/Dialect/Polygeist/Analysis/MemoryAccessAnalysis.cpp:526]: I guess there isn't anything that prevents an operation that implements the `castOpInterface` to have more than a single operand, although that would be strange. I will add an assertion.

**Review by whitneywhtsang — APPROVED** [review]

**Pull request merged**